[vikeedough] iP

vikeedough opened the pull request:
# Dudu :bear:
> "Dudu helps me keep track of my everyday tasks!" - David Vicedo

Here is a list of things that Dudu can help you with:
- **Add** a task
- **View** your current list of tasks
- **Mark** a task as completed
- **Delete** a task

There are three different types of Tasks that Dudu recognizes!
1. _Todo_ Tasks
2. _Deadline_ Tasks (Create a deadline for this task by using the /by keyword!)
3. _Event_ Tasks (Indicate a duration that this task will have by using the /from and /to keywords!)

What Dudu could possibly show you:
- [x] Read up on CS2103T Topics
- [x] CS2103T Tutorial (by: Feb 5 2025)
- [x] CS2103T ip Week 3 (from: Jan 28 2025, 3:00 pm to: Jan 31 2025, 4:00 pm)

Download [Dudu](https://github.com/vikeedough/ip/releases/tag/A-JavaDoc) and try him out yourself!


Here is the `main` method of Dudu:
```java
public static void main(String[] args) {
    Dudu dudu = new Dudu();
    try {
        dudu.run();
    } catch (DuduException | IOException e) {
        System.out.println("Error: " + e.getMessage());
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

vikeedough committed:
Create utils class

vikeedough committed:
Add receiveInput and echo

vikeedough committed:
Add Entry and EntryList classes

vikeedough committed:
Add mark and unmark

vikeedough committed:
Change code structure

vikeedough committed:
Add Todo, Deadline, Event

vikeedough committed:
Add basic tests

vikeedough committed:
Add tests for invalid input

vikeedough committed:
Add DuduException and help command

vikeedough committed:
Add delete command

vikeedough committed:
Add Parse, Command and Ui classes

vikeedough committed:
Add FileOperation

vikeedough committed:
Merge branch 'branch-level-7'

vikeedough committed:
Update duduTasks.txt

vikeedough committed:
Ensure data directory exists

vikeedough committed:
Add DateTimeParser

vikeedough committed:
Merge branch 'branch-level-8'

vikeedough committed:
Refactor Dudu.java

vikeedough committed:
Merge branch 'branch-a-more-oop'

vikeedough committed:
Add dudu package and shift controller, model and utils

vikeedough committed:
Merge branch 'branch-a-packages'

vikeedough committed:
Merge branch 'branch-a-gradle'

vikeedough committed:
Set up gradle in project

vikeedough committed:
Update duduTasks.txt

vikeedough committed:
Merge branch 'branch-a-gradle'

vikeedough committed:
Create test cases for Parser and DateTimeParser

vikeedough committed:
Add test cases for Parser and DateTimeParser

vikeedough committed:
Merge branch 'branch-a-junit'

vikeedough committed:
Create JAR file

vikeedough committed:
Merge branch 'branch-a-jar'

vikeedough committed:
Add JavaDoc comments

vikeedough committed:
Adjust import statements

vikeedough committed:
Add FindCommand

vikeedough committed:
Merge branch 'branch-Level-9'

vikeedough committed:
Merge branch 'branch-A-CodingStandard'

vikeedough committed:
Merge branch 'branch-A-JavaDoc'

vikeedough committed:
Create GUI for Dudu

vikeedough committed:
Enhance GUI and added colors

vikeedough committed:
Add duduTasks.txt to .gitignore

vikeedough committed:
Merge branch 'branch-level-10'

vikeedough committed:
Add title to Stage

keyansheng commented on src/main/java/dudu/controller/FileOperation.java:
```diff
@@ -0,0 +1,152 @@
+        String done = parts[1];
+        String name = parts[2];
+
+        switch (task) {
```
`task` could be renamed to `taskType` since a reader might assume `task` is a `Task` object, which is not the case here.

keyansheng commented on src/main/java/dudu/controller/FileOperation.java:
```diff
@@ -0,0 +1,152 @@
+        }
+
+        String task = parts[0];
+        String done = parts[1];
```
Consider using a boolean `isDone` instead of repeating the same check in every case.
```suggestion
        boolean isDone = parts[1].equals("Done");
```

keyansheng commented on src/main/java/dudu/controller/FileOperation.java:
```diff
@@ -0,0 +1,152 @@
+
+        String isDone = task.getIsDone() ? "Done" : "Not Done";
+
+        if (task instanceof Deadline) {
+            return String.format("%s | %s | %s | %s", taskType, isDone, task.getName(), ((Deadline) task).getBy());
+        } else if (task instanceof Event) {
+            return String.format("%s | %s | %s | %s | %s", taskType, isDone, task.getName(),
+                    ((Event) task).getFrom(), ((Event) task).getTo());
+        } else {
+            return String.format("%s | %s | %s", taskType, isDone, task.getName());
+        }
```
Consider implementing this within each task class for better encapsulation and adherence to the Tell, Don't Ask principle. This would also avoid the need of lots of casting to access subclass methods.

keyansheng reviewed (commented):
Code is well organised and well named overall 👍

reuben-thomas commented on src/main/java/dudu/DialogBox.java:
```diff
@@ -0,0 +1,70 @@
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
+
+import java.io.IOException;
```
The `java.*` imports should be above the others, and between each type of statement, there should be a blank line in between each import type. This is as per one of the guides they shared for [configuring code style in IntelliJ IDEA](https://se-education.org/guides/tutorials/intellijCodeStyle.html).

reuben-thomas commented on src/main/java/dudu/controller/ListCommand.java:
```diff
@@ -0,0 +1,53 @@
+        if (tasksToPrint.isEmpty()) {
+            Ui.printContent("You have no tasks to show.");
+        } else {
+            for(int i=0; i < tasksToPrint.size(); i++) {
```
This is a nitpick, but I believe the style convention mandates a space in `i = 0`

reuben-thomas commented on src/main/java/dudu/Dudu.java:
```diff
@@ -0,0 +1,79 @@
+    private final TaskList tasks;
+    private File cachedTasks;
+
+    /** Creates Dudu object. **/
```
I'm not sure how strict CS2103T is on this, but both the checkstyle and the JavaDoc examples show single-sentence function comments split with the comment open and closed on separate lines.

``` Java
/**
 * Creates Dudu object.
 **/
```

https://www.oracle.com/sg/technical-resources/articles/java/javadoc-tool.html

reuben-thomas commented on src/main/resources/view/DialogBox.fxml:
```diff
@@ -0,0 +1,22 @@
+<?import javafx.scene.image.ImageView?>
+<?import javafx.scene.layout.HBox?>
+
+<fx:root alignment="TOP_RIGHT" maxHeight="1.7976931348623157E308" maxWidth="1.7976931348623157E308" prefWidth="400.0" stylesheets="@../css/dialog-box.css" type="javafx.scene.layout.HBox" xmlns="http://javafx.com/javafx/17" xmlns:fx="http://javafx.com/fxml/1">
```
It might be better to split this into multiple lines.

reuben-thomas reviewed (commented):
Theres just a couple of small convention issues, that should be quick to fix.

1. Order of import statements alphabetically, organized by type separated by a space. You can follow the guide they provided [here](https://se-education.org/guides/tutorials/intellijCodeStyle.html). Unfortunately, a large number of files suffer from this, but others are ok. By chance when you run the IntelliJ formatter, have you been clicking `Only changes uncomited to VCS`? ![image](https://github.com/user-attachments/assets/562b1d40-e81f-44b5-91fd-52abaf8729cb)

2. JavaDoc comment style. Sometimes for function documentation comments, you collapse them into a single line, which doesn't seem to align with the examples the course provided, nor the checkstyle. There are some examples [here](https://www.oracle.com/sg/technical-resources/articles/java/javadoc-tool.html) as well.

Of course, these are very minor nitpicks you can solve in a few minutes! The PR was overall very thorough with descriptive comments for every function. Furthermore I felt the organization was clear and easy to follow for someone looking at your codebase for the first time, so good work!

Wishing you the best!

vikeedough committed:
Edit build.gradle to create jar file from dudu.Launcher

vikeedough committed:
Add assertions

Current code does not have assertions.

Assertions are helpful to to detect issues in the program state at
runtime.

Let's add assertions to identify any possible bugs in the code.

vikeedough committed:
Enhance code quality of Dudu

Codebase is messy and has long methods.

Long methods are hard to read and we may struggle to understand the
code that we are reviewing.

Let's
* avoid complicated expressions
* avoid deep nesting
* shorten length of methods

vikeedough committed:
Merge pull request #1 from vikeedough/branch-A-Assertions

vikeedough committed:
Merge pull request #2 from vikeedough/branch-A-CodeQuality

vikeedough committed:
Add schedule command

vikeedough committed:
Implement schedule command and fix addEntry

vikeedough committed:
Merge branch 'branch-B-ViewSchedules'

vikeedough committed:
Create gradle.yml

vikeedough committed:
Update README.md

vikeedough committed:
Merge branch 'master' of https://github.com/vikeedough/ip

vikeedough committed:
Add product screenshot and clean up error messages

vikeedough committed:
Add schedule to list of commands

vikeedough committed:
Update README.md

vikeedough committed:
Merge branch 'master' of https://github.com/vikeedough/ip

vikeedough committed:
Add README.md to docs folder

vikeedough committed:
Improve code readability in Mark and Unmark Commands

vikeedough committed:
Update README.md

vikeedough committed:
Merge branch 'master' of https://github.com/vikeedough/ip

vikeedough committed:
Handle invalid numbers in marking commands and add delete to help

vikeedough committed:
Edit output text of list command

vikeedough committed:
Add error handling for delete command
